`force=True` is more comprehensive for `powdenest`

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
fixes #19627

#### Brief description of what is fixed or changed

`powdenest` did not always denest exponents; now, except for when the base is negative, the denesting should occur.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* simplify
  * use of `force=True` with `powdenest` is more effective
<!-- END RELEASE NOTES -->

asmeurer commented:
> powdenest did not always denest exponents; now, except for when the base is negative, the denesting should occur.

Can you give an example where it explicitly doesn't denest? Would it be sufficient for force=True to replace checks for `is_assumption is True` with `is_assumption is not False`, so that it disallows cases that are known to not work but doesn't fail on unknown assumptions? Although I'm also partial to the idea that it should just always work regardless of assumptions when force=True.

smichr committed:
powdenest: greater action by force

smichr committed:
watch out for exp_polar

smichr commented on sympy/functions/elementary/complexes.py:
```diff
@@ -530,19 +530,27 @@ def eval(cls, arg):
         from sympy.simplify.simplify import signsimp
         from sympy.core.function import expand_mul
         from sympy.core.power import Pow
+        from sympy.functions.elementary.exponential import exp_polar
```
```suggestion
```

smichr commented on sympy/functions/elementary/complexes.py:
```diff
@@ -530,19 +530,27 @@ def eval(cls, arg):
         if d.free_symbols and not n.free_symbols:
             return cls(n)/cls(d)
 
+        if not arg.atoms(exp_polar):
+            if arg.is_extended_positive:
+                return arg
+            elif arg.is_extended_negative:
+                return -arg
```
```suggestion
```

smichr committed:
Update complexes.py

Don't use `expr.is_extended_positive` since special functions may be so but have a different form after application of `Abs`.

smichr commented:
> it disallows cases that are known to not work but doesn't fail on unknown assumptions

This is still build in as a safety: if the base is negative then no change will be allowed.

And example from the added tests of something that wouldn't denest before is:
```python
>>> powdenest(sqrt(sin(x)**2), force=True) == sin(x)
True
```
This isn't always negative so we allow the user to make this change. But replacing `sin` with `Function('f', negative=True)` would not denest.
